src/Libraries/Microsoft.Extensions.DataIngestion/Chunkers/DocumentTokenChunker.cs

```diff
@@ -0,0 +1,118 @@
+{
+    /// <summary>
+    /// Processes a document by tokenizing its content and dividing it into overlapping chunks of tokens.
+    /// </summary>
+    /// <remarks>
+    /// <para>This class uses a tokenizer to convert the document's content into tokens and then splits the
+    /// tokens into chunks of a specified size, with a configurable overlap between consecutive chunks.</para>
+    /// <para>Note that tables may be split mid-row.</para>
+    /// </remarks>
+    public sealed class DocumentTokenChunker : IngestionChunker<string>
+    {
+        private readonly Tokenizer _tokenizer;
+        private readonly int _maxTokensPerChunk;
+        private readonly int _chunkOverlap;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DocumentTokenChunker"/> class with the specified options.
+        /// </summary>
+        /// <param name="options">The options used to configure the chunker, including tokenizer and chunk sizes.</param>
+        public DocumentTokenChunker(IngestionChunkerOptions options)
+        {
+            _ = Throw.IfNull(options);
+
+            _tokenizer = options.Tokenizer;
+            _maxTokensPerChunk = options.MaxTokensPerChunk;
+            _chunkOverlap = options.OverlapTokens;
+        }
+
+        /// <inheritdoc/>
+        public override async IAsyncEnumerable<IngestionChunk<string>> ProcessAsync(IngestionDocument document, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            _ = Throw.IfNull(document);
+
+            int prevOverlapTokenCount = 0;
+            string? prevOverlap = string.Empty;
+            int stringBuilderTokenCount = 0;
+            StringBuilder stringBuilder = new();
+            foreach (IngestionDocumentElement element in document.EnumerateContent())
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                string? semanticContentToProcess = element.GetSemanticContent();
+                if (string.IsNullOrEmpty(semanticContentToProcess))
+                {
+                    continue;
+                }
+
+                int elementTokenCount = _tokenizer.CountTokens(semanticContentToProcess!, considerNormalization: false);
+                while (prevOverlapTokenCount + stringBuilderTokenCount + elementTokenCount > _maxTokensPerChunk)
+                {
+                    int index = _tokenizer.GetIndexByTokenCount(
+                        text: semanticContentToProcess!,
+                        maxTokenCount: _maxTokensPerChunk - prevOverlapTokenCount - stringBuilderTokenCount,
+                        out string? _,
+                        out int _,
+                        considerNormalization: false);
+
+                    ReadOnlySpan<char> spanToAppend = semanticContentToProcess.AsSpan(0, index);
+#if NET
+                    stringBuilder.Append(spanToAppend);
+#else
+                    stringBuilder.Append(spanToAppend.ToString());
```

Comment:
This could use the overload of Append that takes a pointer and length, allowing you to avoid allocating all of these substrings.